File: txtboard-srv/lib/txtboard/protocol.rb
# txtboard/protocol.rb
#
# Protocolo de wire sobre TCP (torsocks).
# Cada frame é um MessagePack array:
#
#   [ version, type, payload_hash ]
#
# version  : Integer (atualmente 1)
# type     : Symbol serializado como String
# payload  : Hash com campos do frame
#
# Todos os frames que carregam conteúdo de usuário incluem:
#   pubkey  : String hex (64 chars) — chave pública Ed25519 do autor
#   sig     : String hex (128 chars) — assinatura Ed25519 do payload canônico
#
# O receptor SEMPRE verifica a assinatura antes de processar.
# Frames de nós banidos são silenciosamente descartados.

require "msgpack"
require "ed25519"

module Txtboard
  module Protocol
    VERSION = 1

    # ── tipos de frame ────────────────────────────────────────────
    TYPES = %w[
      handshake          # apresentação inicial entre nós
      handshake_ack      # confirmação de handshake
      ping               # keepalive
      pong

      post_announce      # "tenho um post novo em /board/id"
      post_request       # "me manda o post /board/id"
      post_data          # payload completo do post

      image_announce     # "tenho imagem /board/id/filename"
      image_request
      image_chunk        # transferência em chunks (64 KB)
      image_end

      chat_message       # mensagem de relay
      chat_dm            # DM direto entre dois nós

      ban_broadcast      # admin anuncia ban de pubkey
      peer_list          # lista de peers conhecidos
    ].freeze

    # ── construção de frames ──────────────────────────────────────

    def self.build(type, payload = {})
      raise ArgumentError, "tipo desconhecido: #{type}" unless TYPES.include?(type.to_s)
      MessagePack.pack([VERSION, type.to_s, payload])
    end

    # Constrói frame assinado. identity deve responder a #sign(bytes) → hex
    def self.build_signed(type, payload, identity)
      canonical = canonical_bytes(type, payload)
      sig       = identity.sign(canonical)
      build(type, payload.merge("sig" => sig, "pubkey" => identity.pubkey_hex))
    end

    # ── parsing ───────────────────────────────────────────────────

    ParsedFrame = Struct.new(:version, :type, :payload, keyword_init: true)

    def self.parse(raw_bytes)
      arr = MessagePack.unpack(raw_bytes)
      raise ProtocolError, "frame malformado" unless arr.is_a?(Array) && arr.size == 3
      version, type, payload = arr
      raise ProtocolError, "versão incompatível: #{version}" unless version == VERSION
      raise ProtocolError, "tipo inválido: #{type}" unless TYPES.include?(type)
      ParsedFrame.new(version: version, type: type, payload: payload)
    rescue MessagePack::MalformedFormatError => e
      raise ProtocolError, "msgpack inválido: #{e.message}"
    end

    # ── verificação de assinatura ─────────────────────────────────

    def self.verify!(frame)
      p = frame.payload
      pubkey_hex = p["pubkey"] or raise ProtocolError, "sem pubkey"
      sig_hex    = p["sig"]    or raise ProtocolError, "sem sig"

      payload_without_sig = p.reject { |k, _| k == "sig" }
      canonical = canonical_bytes(frame.type, payload_without_sig)

      pk_bytes  = [pubkey_hex].pack("H*")
      sig_bytes = [sig_hex].pack("H*")

      vk  = Ed25519::VerifyKey.new(pk_bytes)
      vk.verify(sig_bytes, canonical)
    rescue Ed25519::VerifyError
      raise SignatureError, "assinatura inválida de #{p["pubkey"]&.slice(0, 12)}"
    end

    # Bytes canônicos para assinatura: "VERSION|type|key1=val1|key2=val2..." (chaves sorted)
    def self.canonical_bytes(type, payload)
      parts = payload
        .reject { |k, _| k == "sig" }
        .sort_by { |k, _| k.to_s }
        .map { |k, v| "#{k}=#{v}" }
      "#{VERSION}|#{type}|#{parts.join("|")}".b
    end

    # ── framing sobre stream TCP ──────────────────────────────────
    # Prefixo de 4 bytes (big-endian uint32) com tamanho do payload.

    def self.encode(raw_frame)
      [raw_frame.bytesize].pack("N") + raw_frame
    end

    # Lê um frame completo de um IO. Bloqueia até ter dados suficientes.
    def self.read_frame(io)
      len_bytes = io.read(4) or raise EOFError
      len = len_bytes.unpack1("N")
      raise ProtocolError, "frame absurdo: #{len} bytes" if len > 4 * 1024 * 1024
      io.read(len) or raise EOFError
    end

    # ── erros ─────────────────────────────────────────────────────
    class ProtocolError  < StandardError; end
    class SignatureError < StandardError; end
  end
end

File: pyon-relay/src/protocol.rs
use std::collections::BTreeMap;

use pyon_core::identity::Identity;

pub const VERSION: u8 = 1;

pub const TYPE_HANDSHAKE: &str = "handshake";
pub const TYPE_HANDSHAKE_ACK: &str = "handshake_ack";
pub const TYPE_PING: &str = "ping";
pub const TYPE_PONG: &str = "pong";
pub const TYPE_POST_ANNOUNCE: &str = "post_announce";
pub const TYPE_POST_REQUEST: &str = "post_request";
pub const TYPE_POST_DATA: &str = "post_data";
pub const TYPE_CHAT_MESSAGE: &str = "chat_message";
pub const TYPE_CHAT_DM: &str = "chat_dm";
pub const TYPE_BAN_BROADCAST: &str = "ban_broadcast";
pub const TYPE_PEER_LIST: &str = "peer_list";

pub const VALID_TYPES: &[&str] = &[
    TYPE_HANDSHAKE,
    TYPE_HANDSHAKE_ACK,
    TYPE_PING,
    TYPE_PONG,
    TYPE_POST_ANNOUNCE,
    TYPE_POST_REQUEST,
    TYPE_POST_DATA,
    TYPE_CHAT_MESSAGE,
    TYPE_CHAT_DM,
    TYPE_BAN_BROADCAST,
    TYPE_PEER_LIST,
];

pub type Payload = BTreeMap<String, String>;

#[derive(Debug, Clone)]
pub struct Frame {
    pub version: u8,
    pub type_: String,
    pub payload: Payload,
}

pub fn build_raw(type_: &str, payload: &Payload) -> Vec<u8> {
    rmp_serde::to_vec(&(VERSION, type_, payload)).unwrap()
}

pub fn encode(msgpack_bytes: &[u8]) -> Vec<u8> {
    let len = (msgpack_bytes.len() as u32).to_be_bytes();
    [&len[..], msgpack_bytes].concat()
}

pub fn parse(data: &[u8]) -> Result<Frame, String> {
    let (version, type_, payload): (u8, String, Payload) =
        rmp_serde::from_slice(data).map_err(|e| format!("msgpack inválido: {}", e))?;
    if version != VERSION {
        return Err(format!("versão incompatível: {}", version));
    }
    if !VALID_TYPES.contains(&type_.as_str()) {
        return Err(format!("tipo inválido: {}", type_));
    }
    Ok(Frame {
        version,
        type_,
        payload,
    })
}

pub fn build(type_: &str, payload: Payload) -> Vec<u8> {
    let raw = build_raw(type_, &payload);
    encode(&raw)
}

pub fn build_signed(type_: &str, mut payload: Payload, identity: &Identity) -> Vec<u8> {
    payload.insert("pubkey".into(), identity.pubkey_hex.clone());
    let canonical = canonical_bytes(type_, &payload);
    let sig = identity.sign(&canonical);
    payload.insert("sig".into(), sig);
    build(type_, payload)
}

pub fn canonical_bytes(type_: &str, payload: &Payload) -> Vec<u8> {
    let parts: Vec<String> = payload
        .iter()
        .filter(|(k, _)| *k != "sig")
        .map(|(k, v)| format!("{}={}", k, v))
        .collect();
    format!("{}|{}|{}", VERSION, type_, parts.join("|")).into_bytes()
}

pub fn verify(frame: &Frame) -> Result<(), String> {
    let pubkey_hex = frame
        .payload
        .get("pubkey")
        .ok_or_else(|| "sem pubkey".to_string())?;
    let sig_hex = frame
        .payload
        .get("sig")
        .ok_or_else(|| "sem sig".to_string())?;

    let payload_without_sig: Payload = frame
        .payload
        .iter()
        .filter(|(k, _)| *k != "sig")
        .map(|(k, v)| (k.clone(), v.clone()))
        .collect();
    let canonical = canonical_bytes(&frame.type_, &payload_without_sig);

    match Identity::verify(pubkey_hex, sig_hex, &canonical) {
        Ok(true) => Ok(()),
        Ok(false) => Err(format!(
            "assinatura inválida de {}",
            &pubkey_hex[..pubkey_hex.len().min(12)]
        )),
        Err(e) => Err(format!("erro de verificação: {}", e)),
    }
}
